{"cells":[{"cell_type":"markdown","metadata":{"id":"b5fMowVXinnt"},"source":["# Deciphering cell-cell communication in single-cell transcriptomics data\n"]},{"cell_type":"markdown","metadata":{"id":"1qOd4r2Siqug"},"source":["## Install the required libraries"]},{"cell_type":"code","execution_count":1,"metadata":{"id":"QJ1QZh_4zV-1","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1698280546814,"user_tz":180,"elapsed":34202,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"9a13b69e-49b7-46b1-de6d-4aaa2385bf0a"},"outputs":[{"output_type":"stream","name":"stderr","text":["Tracing function \"install.packages\" in package \"utils\"\n","\n"]}],"source":["download.file(\"https://github.com/eddelbuettel/r2u/raw/master/inst/scripts/add_cranapt_jammy.sh\",\n","              \"add_cranapt_jammy.sh\")\n","Sys.chmod(\"add_cranapt_jammy.sh\", \"0755\")\n","system(\"./add_cranapt_jammy.sh\")\n","bspm::enable()\n","options(bspm.version.check=FALSE)"]},{"cell_type":"markdown","metadata":{"id":"U2cdonBMivhZ"},"source":["We will create an R function to performs system calls"]},{"cell_type":"code","execution_count":2,"metadata":{"id":"6uBhWJgUiHVH","executionInfo":{"status":"ok","timestamp":1698280546814,"user_tz":180,"elapsed":18,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}}},"outputs":[],"source":["shell_call <- function(command, ...) {\n","  result <- system(command, intern = TRUE, ...)\n","  cat(paste0(result, collapse = \"\\n\"))\n","}"]},{"cell_type":"markdown","metadata":{"id":"I2EVnawvpIa8"},"source":["Install required libraries"]},{"cell_type":"code","execution_count":3,"metadata":{"id":"1EgIyBgfmlfB","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1698280801504,"user_tz":180,"elapsed":254707,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"83d0887c-5d69-4deb-cbfd-5b6671e26988"},"outputs":[{"output_type":"stream","name":"stderr","text":["Downloading GitHub repo satijalab/seurat-wrappers@d28512f804d5fe05e6d68900ca9221020d52cf1d\n","\n","Installing 72 packages: Rcpp, sitmo, BH, dqrng, RcppProgress, irlba, RcppAnnoy, FNN, polyclip, deldir, spatstat.utils, spatstat.data, tensor, abind, goftest, spatstat.sparse, spatstat.random, spatstat.geom, promises, later, sourcetools, xtable, httpuv, parallelly, listenv, globals, RcppEigen, progressr, future.apply, future, sp, plyr, RcppArmadillo, matrixStats, gridExtra, reshape2, bitops, caTools, gtools, gplots, png, here, RcppTOML, crosstalk, lazyeval, htmlwidgets, shiny, zoo, igraph, reticulate, uwot, spatstat.explore, SeuratObject, sctransform, scattermore, Rtsne, ROCR, RANN, plotly, pbapply, patchwork, miniUI, lmtest, leiden, ica, ggridges, ggrepel, fitdistrplus, cowplot, Seurat, rsvd, BiocManager\n","\n"]},{"output_type":"stream","name":"stdout","text":["\u001b[36m──\u001b[39m \u001b[36mR CMD build\u001b[39m \u001b[36m─────────────────────────────────────────────────────────────────\u001b[39m\n","* checking for file ‘/tmp/RtmpWyHo4m/remotes136115a3dcd/satijalab-seurat-wrappers-d28512f/DESCRIPTION’ ... OK\n","* preparing ‘SeuratWrappers’:\n","* checking DESCRIPTION meta-information ... OK\n","* checking for LF line-endings in source and make files and shell scripts\n","* checking for empty or unneeded directories\n","Omitted ‘LazyData’ from DESCRIPTION\n","* building ‘SeuratWrappers_0.3.1.tar.gz’\n","\n"]},{"output_type":"stream","name":"stderr","text":["Installing package into ‘/usr/local/lib/R/site-library’\n","(as ‘lib’ is unspecified)\n","\n","Downloading GitHub repo saezlab/liana@HEAD\n","\n","Downloading GitHub repo saezlab/OmnipathR@HEAD\n","\n","Installing 2 packages: logger, checkmate\n","\n"]},{"output_type":"stream","name":"stdout","text":["\u001b[36m──\u001b[39m \u001b[36mR CMD build\u001b[39m \u001b[36m─────────────────────────────────────────────────────────────────\u001b[39m\n","* checking for file ‘/tmp/RtmpWyHo4m/remotes13617f916fde/saezlab-OmnipathR-577f30d/DESCRIPTION’ ... OK\n","* preparing ‘OmnipathR’:\n","* checking DESCRIPTION meta-information ... OK\n","* checking for LF line-endings in source and make files and shell scripts\n","* checking for empty or unneeded directories\n","* looking to see if a ‘data/datalist’ file should be added\n","* building ‘OmnipathR_3.11.1.tar.gz’\n","\n"]},{"output_type":"stream","name":"stderr","text":["Installing package into ‘/usr/local/lib/R/site-library’\n","(as ‘lib’ is unspecified)\n","\n","Downloading GitHub repo LTLA/basilisk.utils@HEAD\n","\n","Installing 2 packages: filelock, dir.expiry\n","\n"]},{"output_type":"stream","name":"stdout","text":["\u001b[36m──\u001b[39m \u001b[36mR CMD build\u001b[39m \u001b[36m─────────────────────────────────────────────────────────────────\u001b[39m\n","* checking for file ‘/tmp/RtmpWyHo4m/remotes136169542963/LTLA-basilisk.utils-0a17046/DESCRIPTION’ ... OK\n","* preparing ‘basilisk.utils’:\n","* checking DESCRIPTION meta-information ... OK\n","* checking for LF line-endings in source and make files and shell scripts\n","* checking for empty or unneeded directories\n","* building ‘basilisk.utils_1.13.3.tar.gz’\n","\n"]},{"output_type":"stream","name":"stderr","text":["Installing package into ‘/usr/local/lib/R/site-library’\n","(as ‘lib’ is unspecified)\n","\n","Downloading GitHub repo LTLA/basilisk@HEAD\n","\n"]},{"output_type":"stream","name":"stdout","text":["\u001b[36m──\u001b[39m \u001b[36mR CMD build\u001b[39m \u001b[36m─────────────────────────────────────────────────────────────────\u001b[39m\n","* checking for file ‘/tmp/RtmpWyHo4m/remotes13611e57bc7f/LTLA-basilisk-8014ad7/DESCRIPTION’ ... OK\n","* preparing ‘basilisk’:\n","* checking DESCRIPTION meta-information ... OK\n","* checking for LF line-endings in source and make files and shell scripts\n","* checking for empty or unneeded directories\n","* building ‘basilisk_1.13.4.tar.gz’\n","\n"]},{"output_type":"stream","name":"stderr","text":["Installing package into ‘/usr/local/lib/R/site-library’\n","(as ‘lib’ is unspecified)\n","\n","Installing 60 packages: filelock, dir.expiry, iterators, foreach, S4Vectors, BiocGenerics, GlobalOptions, rjson, shape, zlibbioc, XVector, S4Arrays, IRanges, MatrixGenerics, SparseArray, DelayedArray, sparseMatrixStats, GenomeInfoDbData, RCurl, GenomeInfoDb, Biobase, GenomicRanges, formatR, futile.options, lambda.r, snow, futile.logger, SummarizedExperiment, vipor, beeswarm, ggbeeswarm, Cairo, beachmat, ScaledMatrix, BiocParallel, RcppHNSW, DelayedMatrixStats, SingleCellExperiment, BiocNeighbors, statmod, locfit, limma, doParallel, clue, GetoptLong, circlize, ggrastr, pheatmap, RcppML, viridis, BiocSingular, scuttle, metapod, bluster, edgeR, logger, checkmate, ComplexHeatmap, scater, scran\n","\n","Installing package into ‘/usr/local/lib/R/site-library’\n","(as ‘lib’ is unspecified)\n","\n","also installing the dependency ‘S4Arrays’\n","\n","\n","Skipping install of 'OmnipathR' from a github remote, the SHA1 (577f30d9) has not changed since last install.\n","  Use `force = TRUE` to force installation\n","\n","Skipping install of 'basilisk.utils' from a github remote, the SHA1 (0a17046a) has not changed since last install.\n","  Use `force = TRUE` to force installation\n","\n","Skipping install of 'basilisk' from a github remote, the SHA1 (8014ad72) has not changed since last install.\n","  Use `force = TRUE` to force installation\n","\n"]},{"output_type":"stream","name":"stdout","text":["\u001b[36m──\u001b[39m \u001b[36mR CMD build\u001b[39m \u001b[36m─────────────────────────────────────────────────────────────────\u001b[39m\n","* checking for file ‘/tmp/RtmpWyHo4m/remotes1361190928cc/saezlab-liana-86adaca/DESCRIPTION’ ... OK\n","* preparing ‘liana’:\n","* checking DESCRIPTION meta-information ... OK\n","* checking for LF line-endings in source and make files and shell scripts\n","* checking for empty or unneeded directories\n","* building ‘liana_0.1.12.tar.gz’\n","\n"]},{"output_type":"stream","name":"stderr","text":["Installing package into ‘/usr/local/lib/R/site-library’\n","(as ‘lib’ is unspecified)\n","\n"]}],"source":["install.packages(\"R.utils\")\n","remotes::install_github('satijalab/seurat-wrappers@d28512f804d5fe05e6d68900ca9221020d52cf1d', upgrade=F)\n","if (!require(\"BiocManager\", quietly = TRUE))\n","    install.packages(\"BiocManager\", quiet = T)\n","\n","install.packages(\"harmony\")\n","remotes::install_github('saezlab/liana', upgrade=F)"]},{"cell_type":"markdown","metadata":{"id":"vjg2X61cpzdh"},"source":["## Introduction"]},{"cell_type":"markdown","metadata":{"id":"wsg9Zkncp2gF"},"source":["LIANA provides different statistical methods to infer ligand-receptor interactions from single-cell transcriptomics data omics data using prior knowledge. In this notebook we showcase how to use liana in its' most basic form with our data of interest.\n","\n","![LIANA](https://saezlab.github.io/liana/articles/ligrec_pipe.png)"]},{"cell_type":"code","execution_count":4,"metadata":{"id":"XFQkSMjPo1U6","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1698280815349,"user_tz":180,"elapsed":13865,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"2b449186-6a7b-4bba-fa60-ce7ee36cbb3d"},"outputs":[{"output_type":"stream","name":"stderr","text":["── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n","\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.3     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4\n","\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.0\n","\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.4     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n","\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n","\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \n","── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n","\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n","\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n","\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n","\n","Attaching package: ‘magrittr’\n","\n","\n","The following object is masked from ‘package:purrr’:\n","\n","    set_names\n","\n","\n","The following object is masked from ‘package:tidyr’:\n","\n","    extract\n","\n","\n","Warning message:\n","“replacing previous import ‘S4Arrays::read_block’ by ‘DelayedArray::read_block’ when loading ‘SummarizedExperiment’”\n","Attaching SeuratObject\n","\n","‘SeuratObject’ was built under R 4.3.0 but the current version is\n","4.3.1; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n","for R may have changed\n","\n"]}],"source":["library(tidyverse)\n","library(magrittr)\n","library(liana)\n","library(Seurat)"]},{"cell_type":"markdown","metadata":{"id":"7h8NnlBbvWh2"},"source":["Here we can show all the methods from other tools that LIANA implements. Each method infers relevant ligand-receptor interactions relying on different assumptions and each method returns different ligand-receptor scores, typically a pair per method. One score corresponding to the magnitude (strength) of interaction and the other reflecting how specificity of a given interaction to a pair cell identities."]},{"cell_type":"code","execution_count":5,"metadata":{"id":"DNS_nfZ9o2pn","colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"status":"ok","timestamp":1698280815349,"user_tz":180,"elapsed":27,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"44f19950-da82-4dc0-d87c-b0726fa4ee09"},"outputs":[{"output_type":"display_data","data":{"text/html":["<style>\n",".list-inline {list-style: none; margin:0; padding: 0}\n",".list-inline>li {display: inline-block}\n",".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n","</style>\n","<ol class=list-inline><li>'connectome'</li><li>'logfc'</li><li>'natmi'</li><li>'sca'</li><li>'cellphonedb'</li><li>'cytotalk'</li><li>'call_squidpy'</li><li>'call_cellchat'</li><li>'call_connectome'</li><li>'call_sca'</li><li>'call_italk'</li><li>'call_natmi'</li></ol>\n"],"text/markdown":"1. 'connectome'\n2. 'logfc'\n3. 'natmi'\n4. 'sca'\n5. 'cellphonedb'\n6. 'cytotalk'\n7. 'call_squidpy'\n8. 'call_cellchat'\n9. 'call_connectome'\n10. 'call_sca'\n11. 'call_italk'\n12. 'call_natmi'\n\n\n","text/latex":"\\begin{enumerate*}\n\\item 'connectome'\n\\item 'logfc'\n\\item 'natmi'\n\\item 'sca'\n\\item 'cellphonedb'\n\\item 'cytotalk'\n\\item 'call\\_squidpy'\n\\item 'call\\_cellchat'\n\\item 'call\\_connectome'\n\\item 'call\\_sca'\n\\item 'call\\_italk'\n\\item 'call\\_natmi'\n\\end{enumerate*}\n","text/plain":[" [1] \"connectome\"      \"logfc\"           \"natmi\"           \"sca\"            \n"," [5] \"cellphonedb\"     \"cytotalk\"        \"call_squidpy\"    \"call_cellchat\"  \n"," [9] \"call_connectome\" \"call_sca\"        \"call_italk\"      \"call_natmi\"     "]},"metadata":{}}],"source":["show_methods()"]},{"cell_type":"markdown","metadata":{"id":"6a4kJL_QvxyU"},"source":["The different resources of ligand-receptor interactions can be found here. The consensus integrates all the other resources."]},{"cell_type":"code","execution_count":6,"metadata":{"id":"LQdFYIeFqHf2","colab":{"base_uri":"https://localhost:8080/","height":0},"executionInfo":{"status":"ok","timestamp":1698280816446,"user_tz":180,"elapsed":1101,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"cd6e3f74-49bb-4fbc-96e4-9005564fa29c"},"outputs":[{"output_type":"display_data","data":{"text/html":["<style>\n",".list-inline {list-style: none; margin:0; padding: 0}\n",".list-inline>li {display: inline-block}\n",".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n","</style>\n","<ol class=list-inline><li>'Default'</li><li>'Consensus'</li><li>'Baccin2019'</li><li>'CellCall'</li><li>'CellChatDB'</li><li>'Cellinker'</li><li>'CellPhoneDB'</li><li>'CellTalkDB'</li><li>'connectomeDB2020'</li><li>'EMBRACE'</li><li>'Guide2Pharma'</li><li>'HPMR'</li><li>'ICELLNET'</li><li>'iTALK'</li><li>'Kirouac2010'</li><li>'LRdb'</li><li>'Ramilowski2015'</li><li>'OmniPath'</li><li>'MouseConsensus'</li></ol>\n"],"text/markdown":"1. 'Default'\n2. 'Consensus'\n3. 'Baccin2019'\n4. 'CellCall'\n5. 'CellChatDB'\n6. 'Cellinker'\n7. 'CellPhoneDB'\n8. 'CellTalkDB'\n9. 'connectomeDB2020'\n10. 'EMBRACE'\n11. 'Guide2Pharma'\n12. 'HPMR'\n13. 'ICELLNET'\n14. 'iTALK'\n15. 'Kirouac2010'\n16. 'LRdb'\n17. 'Ramilowski2015'\n18. 'OmniPath'\n19. 'MouseConsensus'\n\n\n","text/latex":"\\begin{enumerate*}\n\\item 'Default'\n\\item 'Consensus'\n\\item 'Baccin2019'\n\\item 'CellCall'\n\\item 'CellChatDB'\n\\item 'Cellinker'\n\\item 'CellPhoneDB'\n\\item 'CellTalkDB'\n\\item 'connectomeDB2020'\n\\item 'EMBRACE'\n\\item 'Guide2Pharma'\n\\item 'HPMR'\n\\item 'ICELLNET'\n\\item 'iTALK'\n\\item 'Kirouac2010'\n\\item 'LRdb'\n\\item 'Ramilowski2015'\n\\item 'OmniPath'\n\\item 'MouseConsensus'\n\\end{enumerate*}\n","text/plain":[" [1] \"Default\"          \"Consensus\"        \"Baccin2019\"       \"CellCall\"        \n"," [5] \"CellChatDB\"       \"Cellinker\"        \"CellPhoneDB\"      \"CellTalkDB\"      \n"," [9] \"connectomeDB2020\" \"EMBRACE\"          \"Guide2Pharma\"     \"HPMR\"            \n","[13] \"ICELLNET\"         \"iTALK\"            \"Kirouac2010\"      \"LRdb\"            \n","[17] \"Ramilowski2015\"   \"OmniPath\"         \"MouseConsensus\"  "]},"metadata":{}}],"source":["show_resources()"]},{"cell_type":"markdown","metadata":{"id":"imGhC2zTz1F1"},"source":["## Loading Data"]},{"cell_type":"markdown","metadata":{"id":"v5fMnPapoqol"},"source":["Here we load our data of interest to study cell-cell communicaton."]},{"cell_type":"code","source":["download.file(\"https://www.dropbox.com/scl/fi/1ysew52kr8o2riahzubcw/BALF-COVID19-Liao_et_al-NatMed-2020.rds?rlkey=tg3tpn8la6oth25wvx3a22qt9&dl=1\", \"COVID.rds\")"],"metadata":{"id":"snNnQfejYsbO","executionInfo":{"status":"ok","timestamp":1698280822583,"user_tz":180,"elapsed":6141,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}}},"execution_count":7,"outputs":[]},{"cell_type":"code","source":["testdata <- readRDS('COVID.rds')"],"metadata":{"id":"lgWJAv3zw_OQ","executionInfo":{"status":"ok","timestamp":1698280844099,"user_tz":180,"elapsed":21537,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}}},"execution_count":8,"outputs":[]},{"cell_type":"code","execution_count":9,"metadata":{"id":"oXcrZfhUz7nW","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1698280844099,"user_tz":180,"elapsed":25,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"4f3a82d7-af98-4546-dcb3-7bf3ea8ebc91"},"outputs":[{"output_type":"stream","name":"stdout","text":["Formal class 'Seurat' [package \"SeuratObject\"] with 13 slots\n","  ..@ assays      :List of 1\n","  .. ..$ RNA:Formal class 'Assay' [package \"SeuratObject\"] with 8 slots\n","  ..@ meta.data   :'data.frame':\t63103 obs. of  8 variables:\n","  .. ..$ sample    : Factor w/ 12 levels \"C51\",\"C52\",\"C100\",..: 3 3 3 3 3 3 3 3 3 3 ...\n","  .. ..$ sample_new: Factor w/ 12 levels \"HC1\",\"HC2\",\"HC3\",..: 3 3 3 3 3 3 3 3 3 3 ...\n","  .. ..$ group     : Factor w/ 3 levels \"HC\",\"M\",\"S\": 1 1 1 1 1 1 1 1 1 1 ...\n","  .. ..$ disease   : Factor w/ 2 levels \"N\",\"Y\": 1 1 1 1 1 1 1 1 1 1 ...\n","  .. ..$ hasnCoV   : Factor w/ 2 levels \"N\",\"Y\": 1 1 1 1 1 1 1 1 1 1 ...\n","  .. ..$ cluster   : num [1:63103] 27 23 6 10 10 9 0 17 10 10 ...\n","  .. ..$ celltype  : Factor w/ 10 levels \"B\",\"Epithelial\",..: 1 3 8 3 3 8 3 5 3 3 ...\n","  .. ..$ condition : Factor w/ 3 levels \"Control\",\"Moderate COVID-19\",..: 1 1 1 1 1 1 1 1 1 1 ...\n","  ..@ active.assay: chr \"RNA\"\n","  ..@ active.ident: Factor w/ 1 level \"AnnData\": 1 1 1 1 1 1 1 1 1 1 ...\n","  .. ..- attr(*, \"names\")= chr [1:63103] \"AAACCCACAGCTACAT_3\" \"AAACCCATCCACGGGT_3\" \"AAACCCATCCCATTCG_3\" \"AAACGAACAAACAGGC_3\" ...\n","  ..@ graphs      : list()\n","  ..@ neighbors   : list()\n","  ..@ reductions  : list()\n","  ..@ images      : list()\n","  ..@ project.name: chr \"AnnData\"\n","  ..@ misc        : Named list()\n","  ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1\n","  .. ..$ : int [1:4] 3 1 5 9900\n","  ..@ commands    : list()\n","  ..@ tools       : Named list()\n"]}],"source":["testdata %>% dplyr::glimpse()"]},{"cell_type":"code","execution_count":10,"metadata":{"id":"urTCplRfgPG8","executionInfo":{"status":"ok","timestamp":1698280851672,"user_tz":180,"elapsed":7594,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}}},"outputs":[],"source":["testdata <- subset(x = testdata, subset = group == \"S\")"]},{"cell_type":"code","execution_count":11,"metadata":{"id":"xl9GFjY4inBA","executionInfo":{"status":"ok","timestamp":1698280851672,"user_tz":180,"elapsed":22,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}}},"outputs":[],"source":["Idents(testdata) <- \"celltype\""]},{"cell_type":"code","execution_count":12,"metadata":{"id":"4r-ZpqPGi7Ts","executionInfo":{"status":"ok","timestamp":1698280857978,"user_tz":180,"elapsed":6328,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}}},"outputs":[],"source":["testdata <- Seurat::NormalizeData(testdata, verbose = FALSE)"]},{"cell_type":"code","source":["testdata %>% dplyr::glimpse()\n"],"metadata":{"id":"i67x4qgb5U8u","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1698280857978,"user_tz":180,"elapsed":28,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"3ac118bb-346e-4e30-f088-1fd11e4d7c60"},"execution_count":13,"outputs":[{"output_type":"stream","name":"stdout","text":["Formal class 'Seurat' [package \"SeuratObject\"] with 13 slots\n","  ..@ assays      :List of 1\n","  .. ..$ RNA:Formal class 'Assay' [package \"SeuratObject\"] with 8 slots\n","  ..@ meta.data   :'data.frame':\t36619 obs. of  10 variables:\n","  .. ..$ sample      : Factor w/ 12 levels \"C51\",\"C52\",\"C100\",..: 11 11 11 11 11 11 11 11 11 11 ...\n","  .. ..$ sample_new  : Factor w/ 12 levels \"HC1\",\"HC2\",\"HC3\",..: 11 11 11 11 11 11 11 11 11 11 ...\n","  .. ..$ group       : Factor w/ 3 levels \"HC\",\"M\",\"S\": 3 3 3 3 3 3 3 3 3 3 ...\n","  .. ..$ disease     : Factor w/ 2 levels \"N\",\"Y\": 2 2 2 2 2 2 2 2 2 2 ...\n","  .. ..$ hasnCoV     : Factor w/ 2 levels \"N\",\"Y\": 1 1 1 1 1 1 1 1 1 1 ...\n","  .. ..$ cluster     : num [1:36619] 9 6 1 17 11 8 9 0 20 4 ...\n","  .. ..$ celltype    : Factor w/ 10 levels \"B\",\"Epithelial\",..: 8 8 3 5 3 3 8 3 9 3 ...\n","  .. ..$ condition   : Factor w/ 3 levels \"Control\",\"Moderate COVID-19\",..: 3 3 3 3 3 3 3 3 3 3 ...\n","  .. ..$ nCount_RNA  : num [1:36619] 3070 2524 4585 2387 8730 ...\n","  .. ..$ nFeature_RNA: int [1:36619] 1493 1182 1666 1273 2457 1808 1003 1640 511 2927 ...\n","  ..@ active.assay: chr \"RNA\"\n","  ..@ active.ident: Factor w/ 10 levels \"B\",\"Epithelial\",..: 8 8 3 5 3 3 8 3 9 3 ...\n","  .. ..- attr(*, \"names\")= chr [1:36619] \"AAACCTGCAGTGGGAT_12\" \"AAACCTGCATTCGACA_12\" \"AAACCTGTCACCTCGT_12\" \"AAACGGGAGCACCGCT_12\" ...\n","  ..@ graphs      : list()\n","  ..@ neighbors   : list()\n","  ..@ reductions  : list()\n","  ..@ images      : list()\n","  ..@ project.name: chr \"AnnData\"\n","  ..@ misc        : Named list()\n","  ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1\n","  .. ..$ : int [1:4] 3 1 5 9900\n","  ..@ commands    :List of 1\n","  .. ..$ NormalizeData.RNA:Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n","  ..@ tools       : Named list()\n"]}]},{"cell_type":"markdown","metadata":{"id":"7Lt2bhPcoxWZ"},"source":["## Running LIANA"]},{"cell_type":"markdown","metadata":{"id":"SJP0H_1AozzB"},"source":["We can run LIANA by using any of the methods that it implements. In this example, we use the implementation of [CellPhoneDB](https://www.nature.com/articles/s41596-020-0292-x). However, `liana_wrap` calls a number of methods and each method is run with the provided resource(s). If no method is specified, it runs all methods implemented in LIANA. Similarly, the `consensus` is the default resource."]},{"cell_type":"code","execution_count":14,"metadata":{"id":"vV2jImJY0zMU","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1698280935031,"user_tz":180,"elapsed":77077,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"ba201d5c-18a7-4a2d-f79c-6274cdee6b62"},"outputs":[{"output_type":"stream","name":"stderr","text":["Expression from the `RNA` assay will be used\n","\n","Running LIANA with `celltype` as labels!\n","\n","Warning message in exec(output, ...):\n","“6808 genes and/or 0 cells were removed as they had no counts!”\n","Warning message:\n","“\u001b[1m\u001b[22m`invoke()` is deprecated as of rlang 0.4.0.\n","Please use `exec()` or `inject()` instead.\n","\u001b[90mThis warning is displayed once every 8 hours.\u001b[39m”\n","LIANA: LR summary stats calculated!\n","\n","Now Running: Cellphonedb\n","\n","Warning message:\n","“\u001b[1m\u001b[22m`progress_estimated()` was deprecated in dplyr 1.0.0.\n","\u001b[36mℹ\u001b[39m The deprecated feature was likely used in the \u001b[34mliana\u001b[39m package.\n","  Please report the issue at \u001b[3m\u001b[34m<https://github.com/saezlab/liana/issues>\u001b[39m\u001b[23m.”\n"]}],"source":["cpdb_result <- liana_wrap(testdata,\n","                          method = 'cellphonedb',\n","                          resource = c('CellPhoneDB'),\n","                          permutation.params = list(nperms=100,\n","                                                    parallelize=FALSE,\n","                                                    workers=4),\n","                          expr_prop=0.05)"]},{"cell_type":"code","execution_count":15,"metadata":{"id":"lSlH1DNS2VXd","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1698280935032,"user_tz":180,"elapsed":29,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"e3ed5486-604f-407d-a6e6-a7d396ba09a1"},"outputs":[{"output_type":"stream","name":"stdout","text":["Rows: 7,835\n","Columns: 12\n","$ source           \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \"T\", \"T\", \"T\", \"T\", \"T\", \"T\", \"T\", \"T\", \"T\", \"T\", \"T\"…\n","$ target           \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \"T\", \"T\", \"T\", \"T\", \"T\", \"T\", \"T\", \"T\", \"T\", \"T\", \"T\"…\n","$ ligand.complex   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \"CXCL16\", \"CCL5\", \"CCL8\", \"CCL4\", \"LTA\", \"CXCL10\", \"C…\n","$ ligand           \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \"CXCL16\", \"CCL5\", \"CCL8\", \"CCL4\", \"LTA\", \"CXCL10\", \"C…\n","$ receptor.complex \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \"CXCR6\", \"CCR5\", \"CCR5\", \"CCR5\", \"TNFRSF1A\", \"CXCR3\",…\n","$ receptor         \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \"CXCR6\", \"CCR5\", \"CCR5\", \"CCR5\", \"TNFRSF1A\", \"CXCR3\",…\n","$ receptor.prop    \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 0.23829201, 0.24173554, 0.24173554, 0.24173554, 0.164…\n","$ ligand.prop      \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 0.08930211, 0.67791552, 0.22933884, 0.52249770, 0.135…\n","$ ligand.expr      \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 0.10750502, 2.16021901, 0.41310354, 1.41548671, 0.175…\n","$ receptor.expr    \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 0.38005758, 0.28682271, 0.28682271, 0.28682271, 0.174…\n","$ lr.mean          \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 0.2437813, 1.2235209, 0.3499631, 0.8511547, 0.1746789…\n","$ pvalue           \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 1.00, 0.00, 1.00, 1.00, 1.00, 1.00, 0.00, 0.00, 1.00,…\n"]}],"source":["dplyr::glimpse(cpdb_result)"]},{"cell_type":"markdown","metadata":{"id":"V-J3YkxWqdkr"},"source":["We can also select multiple methods, an run all of them. Here we use CellPhoneDB, NATMI, SingleCellSignalR (sca), and the logFC approach."]},{"cell_type":"code","execution_count":16,"metadata":{"id":"3L5UsQzk1mhC","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1698281184258,"user_tz":180,"elapsed":249250,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"8a9cb48f-c7cc-43ab-f812-a2d63003d93c"},"outputs":[{"output_type":"stream","name":"stderr","text":["Expression from the `RNA` assay will be used\n","\n","Running LIANA with `celltype` as labels!\n","\n","Warning message in exec(output, ...):\n","“6808 genes and/or 0 cells were removed as they had no counts!”\n","LIANA: LR summary stats calculated!\n","\n","Now Running: Cellphonedb\n","\n","Now Running: Natmi\n","\n","Now Running: Sca\n","\n","Now Running: Logfc\n","\n"]}],"source":["complex_test <- liana_wrap(testdata,\n","                           method = c('cellphonedb', 'natmi', 'sca', 'logfc'),\n","                           resource = c('CellPhoneDB'))"]},{"cell_type":"code","source":["dplyr::glimpse(complex_test)"],"metadata":{"id":"4O75ssZn_wH-","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1698281184661,"user_tz":180,"elapsed":428,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"2896ae62-4ce5-4b25-9781-4e6048073266"},"execution_count":17,"outputs":[{"output_type":"stream","name":"stdout","text":["List of 4\n"," $ cellphonedb: tibble [4,804 × 12] (S3: tbl_df/tbl/data.frame)\n","  ..$ source          : chr [1:4804] \"T\" \"T\" \"T\" \"T\" ...\n","  ..$ target          : chr [1:4804] \"T\" \"T\" \"T\" \"T\" ...\n","  ..$ ligand.complex  : chr [1:4804] \"CCL5\" \"CCL8\" \"CCL4\" \"LTA\" ...\n","  ..$ ligand          : chr [1:4804] \"CCL5\" \"CCL8\" \"CCL4\" \"LTA\" ...\n","  ..$ receptor.complex: chr [1:4804] \"CCR5\" \"CCR5\" \"CCR5\" \"TNFRSF1A\" ...\n","  ..$ receptor        : chr [1:4804] \"CCR5\" \"CCR5\" \"CCR5\" \"TNFRSF1A\" ...\n","  ..$ receptor.prop   : num [1:4804] 0.242 0.242 0.242 0.164 0.313 ...\n","  ..$ ligand.prop     : num [1:4804] 0.678 0.229 0.522 0.136 0.301 ...\n","  ..$ ligand.expr     : num [1:4804] 2.16 0.413 1.415 0.175 0.596 ...\n","  ..$ receptor.expr   : num [1:4804] 0.287 0.287 0.287 0.174 0.411 ...\n","  ..$ lr.mean         : num [1:4804] 1.224 0.35 0.851 0.175 0.504 ...\n","  ..$ pvalue          : num [1:4804] 0 1 1 1 1 0 0 1 1 1 ...\n"," $ natmi      : tibble [4,804 × 14] (S3: tbl_df/tbl/data.frame)\n","  ..$ source          : chr [1:4804] \"T\" \"T\" \"T\" \"T\" ...\n","  ..$ target          : chr [1:4804] \"T\" \"T\" \"T\" \"T\" ...\n","  ..$ ligand.complex  : chr [1:4804] \"CCL5\" \"CCL8\" \"CCL4\" \"LTA\" ...\n","  ..$ ligand          : chr [1:4804] \"CCL5\" \"CCL8\" \"CCL4\" \"LTA\" ...\n","  ..$ receptor.complex: chr [1:4804] \"CCR5\" \"CCR5\" \"CCR5\" \"TNFRSF1A\" ...\n","  ..$ receptor        : chr [1:4804] \"CCR5\" \"CCR5\" \"CCR5\" \"TNFRSF1A\" ...\n","  ..$ receptor.prop   : num [1:4804] 0.242 0.242 0.242 0.164 0.313 ...\n","  ..$ ligand.prop     : num [1:4804] 0.678 0.229 0.522 0.136 0.301 ...\n","  ..$ ligand.expr     : num [1:4804] 2.16 0.413 1.415 0.175 0.596 ...\n","  ..$ receptor.expr   : num [1:4804] 0.287 0.287 0.287 0.174 0.411 ...\n","  ..$ ligand.sum      : num [1:4804] 6.35 8.121 11.89 0.445 11.484 ...\n","  ..$ receptor.sum    : num [1:4804] 1.32 1.32 1.32 2.89 1.95 ...\n","  ..$ prod_weight     : num [1:4804] 0.6196 0.1185 0.406 0.0305 0.2451 ...\n","  ..$ edge_specificity: num [1:4804] 0.0737 0.011 0.0258 0.0237 0.011 ...\n"," $ sca        : tibble [4,804 × 12] (S3: tbl_df/tbl/data.frame)\n","  ..$ source          : chr [1:4804] \"T\" \"T\" \"T\" \"T\" ...\n","  ..$ target          : chr [1:4804] \"T\" \"T\" \"T\" \"T\" ...\n","  ..$ ligand.complex  : chr [1:4804] \"CCL5\" \"CCL8\" \"CCL4\" \"LTA\" ...\n","  ..$ ligand          : chr [1:4804] \"CCL5\" \"CCL8\" \"CCL4\" \"LTA\" ...\n","  ..$ receptor.complex: chr [1:4804] \"CCR5\" \"CCR5\" \"CCR5\" \"TNFRSF1A\" ...\n","  ..$ receptor        : chr [1:4804] \"CCR5\" \"CCR5\" \"CCR5\" \"TNFRSF1A\" ...\n","  ..$ receptor.prop   : num [1:4804] 0.242 0.242 0.242 0.164 0.313 ...\n","  ..$ ligand.prop     : num [1:4804] 0.678 0.229 0.522 0.136 0.301 ...\n","  ..$ ligand.expr     : num [1:4804] 2.16 0.413 1.415 0.175 0.596 ...\n","  ..$ receptor.expr   : num [1:4804] 0.287 0.287 0.287 0.174 0.411 ...\n","  ..$ global_mean     : num [1:4804] 0.0895 0.0895 0.0895 0.0895 0.0895 ...\n","  ..$ LRscore         : num [1:4804] 0.898 0.794 0.877 0.661 0.847 ...\n"," $ logfc      : tibble [4,804 × 11] (S3: tbl_df/tbl/data.frame)\n","  ..$ source          : chr [1:4804] \"T\" \"T\" \"T\" \"T\" ...\n","  ..$ target          : chr [1:4804] \"T\" \"T\" \"T\" \"T\" ...\n","  ..$ ligand.complex  : chr [1:4804] \"CCL5\" \"CCL8\" \"CCL4\" \"LTA\" ...\n","  ..$ ligand          : chr [1:4804] \"CCL5\" \"CCL8\" \"CCL4\" \"LTA\" ...\n","  ..$ receptor.complex: chr [1:4804] \"CCR5\" \"CCR5\" \"CCR5\" \"TNFRSF1A\" ...\n","  ..$ receptor        : chr [1:4804] \"CCR5\" \"CCR5\" \"CCR5\" \"TNFRSF1A\" ...\n","  ..$ receptor.prop   : num [1:4804] 0.242 0.242 0.242 0.164 0.313 ...\n","  ..$ ligand.prop     : num [1:4804] 0.678 0.229 0.522 0.136 0.301 ...\n","  ..$ ligand.log2FC   : num [1:4804] 3.247 -3.534 -0.447 0.555 -3.112 ...\n","  ..$ receptor.log2FC : num [1:4804] 0.0496 0.0496 0.0496 -0.8804 0.8929 ...\n","  ..$ logfc_comb      : num [1:4804] 1.648 -1.742 -0.199 -0.163 -1.109 ...\n"]}]},{"cell_type":"markdown","metadata":{"id":"AuQjAf3lqyyc"},"source":["One of the key features of LIANA is that it can compute a consensus ranking of the prediction of all methods employed to analyze cell-cell communication. By using the function `liana_aggregate()` we can use all the results from every method used in the previous step."]},{"cell_type":"code","execution_count":18,"metadata":{"id":"hfxOFoyQ1x9s","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1698281185256,"user_tz":180,"elapsed":599,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"1b28877d-3d4a-455a-d39b-41e341834457"},"outputs":[{"output_type":"stream","name":"stderr","text":["Now aggregating cellphonedb\n","\n","Now aggregating natmi\n","\n","Now aggregating sca\n","\n","Now aggregating logfc\n","\n","Aggregating Ranks\n","\n"]}],"source":["liana_consensus <- complex_test %>% liana_aggregate()"]},{"cell_type":"code","execution_count":19,"metadata":{"id":"ZEf8fKf216zY","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1698281185256,"user_tz":180,"elapsed":3,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"5d3d550d-fcef-425d-9bf4-be1a6457c26e"},"outputs":[{"output_type":"stream","name":"stdout","text":["Rows: 4,804\n","Columns: 14\n","$ source                 \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \"T\", \"Macrophages\", \"T\", \"Epithelial\", \"Epithel…\n","$ target                 \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \"T\", \"Macrophages\", \"pDC\", \"Epithelial\", \"Epith…\n","$ ligand.complex         \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \"LCK\", \"CCL7\", \"BST2\", \"SCGB3A1\", \"CEACAM6\", \"H…\n","$ receptor.complex       \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \"CD8A_CD8B\", \"CCR1\", \"LILRA4\", \"NOTCH3\", \"CEACA…\n","$ aggregate_rank         \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 9.869193e-07, 4.233019e-05, 4.485669e-05, 5.085…\n","$ mean_rank              \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 399.50, 317.50, 379.75, 716.00, 322.50, 348.25,…\n","$ cellphonedb.pvalue     \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…\n","$ cellphonedb.rank       \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 808, 808, 808, 808, 808, 808, 808, 808, 808, 80…\n","$ natmi.edge_specificity \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 0.46688113, 0.10625162, 0.10575729, 0.67226981,…\n","$ natmi.rank             \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 15, 274, 278, 7, 1, 436, 341, 90, 371, 375, 398…\n","$ sca.LRscore            \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 0.9046229, 0.9465536, 0.9352749, 0.8453234, 0.9…\n","$ sca.rank               \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 756, 175, 248, 2047, 474, 78, 139, 631, 95, 147…\n","$ logfc.logfc_comb       \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 1.987921, 2.151729, 1.233319, 2.978505, 2.40344…\n","$ logfc.rank             \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m 19, 13, 185, 2, 7, 71, 9, 30, 10, 22, 60, 148, …\n"]}],"source":["dplyr::glimpse(liana_consensus)"]},{"cell_type":"markdown","metadata":{"id":"Lt6tjyZlrF98"},"source":["## Visualization and Interpretation"]},{"cell_type":"markdown","metadata":{"id":"dY27ELgPshLw"},"source":["Dotplots can be generated to easily interpret important ligand-receptor pairs used by sender-receiver cell pairs.\n","\n","Here, we preprocess the results of CellPhoneDB, then plot them. A filter to use only significant cases is applied (P-value < 0.05)."]},{"cell_type":"code","execution_count":20,"metadata":{"id":"zSfhX5732TcT","executionInfo":{"status":"ok","timestamp":1698281185256,"user_tz":180,"elapsed":2,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}}},"outputs":[],"source":["cpdb_int <- cpdb_result %>%\n","  # only keep interactions with p-val <= 0.05\n","  filter(pvalue <= 0.05) %>% # this reflects interactions `specificity`\n","  # then rank according to `magnitude` (lr_mean in this case)\n","  rank_method(method_name = \"cellphonedb\",\n","              mode = \"magnitude\") %>%\n","  # keep top 20 interactions (regardless of cell type)\n","  distinct_at(c(\"ligand.complex\", \"receptor.complex\")) %>%\n","  head(20)"]},{"cell_type":"code","execution_count":21,"metadata":{"id":"Mjox5wT62h3s","executionInfo":{"status":"ok","timestamp":1698281189025,"user_tz":180,"elapsed":3771,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}}},"outputs":[],"source":["#options(repr.plot.height = 12, repr.plot.width = 9)\n","# Plot toy results\n","scPlot <- cpdb_result %>%\n","          # keep only the interactions of interest\n","          inner_join(cpdb_int,\n","                    by = c(\"ligand.complex\", \"receptor.complex\")) %>%\n","          # invert size (low p-value/high specificity = larger dot size)\n","          # + add a small value to avoid Infinity for 0s\n","          mutate(pvalue = -log10(pvalue + 1e-10)) %>%\n","          liana_dotplot(source_groups = c(\"Epithelial\"),\n","                        target_groups = c(\"Macrophages\", \"NK\", \"B\", \"T\", \"Neutrophil\"),\n","                        specificity = \"pvalue\",\n","                        magnitude = \"lr.mean\",\n","                        show_complex = TRUE,\n","                        size.label = \"-log10(p-value)\") + theme(axis.text.x = element_text(angle = 90))\n","ggsave(\"01-liana_dotplot.png\", plot = scPlot, bg = \"white\", dpi = 600, width = 16, height = 9)"]},{"cell_type":"markdown","metadata":{"id":"-xF_p1cbs0Hm"},"source":["Similarly, we can explore the consensus results."]},{"cell_type":"code","execution_count":22,"metadata":{"id":"PrEnrSIi2zCL","executionInfo":{"status":"ok","timestamp":1698281191793,"user_tz":180,"elapsed":2780,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}}},"outputs":[],"source":["#options(repr.plot.height = 12, repr.plot.width = 9)\n","scPlot <- liana_consensus %>%\n","          liana_dotplot(source_groups = c(\"Macrophages\"),\n","                        target_groups = c(\"Macrophages\", \"NK\", \"B\", \"T\", \"Neutrophil\"),\n","                        ntop = 20) + theme(axis.text.x = element_text(angle = 90))\n","ggsave(\"02-liana_dotplot.png\", plot = scPlot, bg = \"white\", dpi = 600, width = 16, height = 9)"]},{"cell_type":"markdown","metadata":{"id":"4EC8pxgHt_6-"},"source":["Overall potential of cells to communicate can be computed. Here, we can count the number of significant/important interactions. Then, they can be visualized through a heatmap.\n","\n","Similarly, we can compare the CellPhoneDB vs consensus results."]},{"cell_type":"code","execution_count":23,"metadata":{"id":"WfmEjOUv27cF","colab":{"base_uri":"https://localhost:8080/","height":34},"executionInfo":{"status":"ok","timestamp":1698281193242,"user_tz":180,"elapsed":1452,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"33022aad-7020-4d33-ae6d-79098f47f04e"},"outputs":[{"output_type":"display_data","data":{"text/html":["<strong>png:</strong> 2"],"text/markdown":"**png:** 2","text/latex":"\\textbf{png:} 2","text/plain":["png \n","  2 "]},"metadata":{}}],"source":["liana_trunc <- cpdb_result %>%\n","   # only keep interactions concordant between methods\n","  filter(pvalue <= 0.05) # note that these pvals are already corrected\n","png(\"03-heat_freq.png\", bg = \"white\")\n","heat_freq(liana_trunc)\n","dev.off()"]},{"cell_type":"code","execution_count":24,"metadata":{"id":"VplzUlt23_zM","colab":{"base_uri":"https://localhost:8080/","height":34},"executionInfo":{"status":"ok","timestamp":1698281196768,"user_tz":180,"elapsed":3530,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"617c048f-8aac-4615-a339-a7d13c76098f"},"outputs":[{"output_type":"display_data","data":{"text/html":["<strong>png:</strong> 2"],"text/markdown":"**png:** 2","text/latex":"\\textbf{png:} 2","text/plain":["png \n","  2 "]},"metadata":{}}],"source":["liana_trunc <- liana_consensus %>%\n","   # only keep interactions concordant between methods\n","  filter(aggregate_rank <= 0.01) # note that these pvals are already corrected\n","png(\"04-heat_freq.png\", bg = \"white\")\n","heat_freq(liana_trunc)\n","dev.off()"]},{"cell_type":"markdown","metadata":{"id":"3yF7ucbruY_j"},"source":["## Extra Questions:\n","\n","- How different are the results between two methods of your choice?\n","\n","- Why can this happen?\n"]},{"cell_type":"code","execution_count":25,"metadata":{"id":"cJoCML_XsUEg","colab":{"base_uri":"https://localhost:8080/","height":225},"executionInfo":{"status":"ok","timestamp":1698281196768,"user_tz":180,"elapsed":8,"user":{"displayName":"Orr Ashenberg","userId":"02232766130052900223"}},"outputId":"97d043a2-f1cb-4a48-f1bc-89f9704ad30e"},"outputs":[{"output_type":"display_data","data":{"text/html":["<style>\n",".list-inline {list-style: none; margin:0; padding: 0}\n",".list-inline>li {display: inline-block}\n",".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n","</style>\n","<ol class=list-inline><li>'assign_lr_weights'</li><li>'calculate_gini'</li><li>'call_cellchat'</li><li>'call_connectome'</li><li>'call_italk'</li><li>'call_natmi'</li><li>'call_sca'</li><li>'call_squidpy'</li><li>'cellchat_formatDB'</li><li>'chord_freq'</li><li>'compile_ligrec'</li><li>'conn_formatDB'</li><li>'decomplexify'</li><li>'decompose_tensor'</li><li>'filter_nonabundant_celltypes'</li><li>'FormatConnectome'</li><li>'FormatiTALK'</li><li>'FormatSCA'</li><li>'generate_homologs'</li><li>'generate_lr_geneset'</li><li>'generate_omni'</li><li>'generate_orthologs'</li><li>'get_abundance_summary'</li><li>'get_c2c_factors'</li><li>'get_connectome'</li><li>'get_curated_omni'</li><li>'get_logfc'</li><li>'get_lr_resources'</li><li>'get_natmi'</li><li>'get_sca'</li><li>'heat_freq'</li><li>'italk_formatDB'</li><li>'join_means'</li><li>'liana_aggregate'</li><li>'liana_bysample'</li><li>'liana_call'</li><li>'liana_defaults'</li><li>'liana_dotplot'</li><li>'liana_heatmap'</li><li>'liana_pipe'</li><li>'liana_prep'</li><li>'liana_tensor_c2c'</li><li>'liana_wrap'</li><li>'mean0'</li><li>'minmax'</li><li>'plot_abundance_summary'</li><li>'plot_c2c_cells'</li><li>'plot_c2c_overview'</li><li>'plot_context_boxplot'</li><li>'plot_context_heat'</li><li>'plot_lr_heatmap'</li><li>'rank_aggregate'</li><li>'rank_method'</li><li>'recomplexify'</li><li>'sca_formatDB'</li><li>'select_resource'</li><li>'show_homologene'</li><li>'show_methods'</li><li>'show_resources'</li></ol>\n"],"text/markdown":"1. 'assign_lr_weights'\n2. 'calculate_gini'\n3. 'call_cellchat'\n4. 'call_connectome'\n5. 'call_italk'\n6. 'call_natmi'\n7. 'call_sca'\n8. 'call_squidpy'\n9. 'cellchat_formatDB'\n10. 'chord_freq'\n11. 'compile_ligrec'\n12. 'conn_formatDB'\n13. 'decomplexify'\n14. 'decompose_tensor'\n15. 'filter_nonabundant_celltypes'\n16. 'FormatConnectome'\n17. 'FormatiTALK'\n18. 'FormatSCA'\n19. 'generate_homologs'\n20. 'generate_lr_geneset'\n21. 'generate_omni'\n22. 'generate_orthologs'\n23. 'get_abundance_summary'\n24. 'get_c2c_factors'\n25. 'get_connectome'\n26. 'get_curated_omni'\n27. 'get_logfc'\n28. 'get_lr_resources'\n29. 'get_natmi'\n30. 'get_sca'\n31. 'heat_freq'\n32. 'italk_formatDB'\n33. 'join_means'\n34. 'liana_aggregate'\n35. 'liana_bysample'\n36. 'liana_call'\n37. 'liana_defaults'\n38. 'liana_dotplot'\n39. 'liana_heatmap'\n40. 'liana_pipe'\n41. 'liana_prep'\n42. 'liana_tensor_c2c'\n43. 'liana_wrap'\n44. 'mean0'\n45. 'minmax'\n46. 'plot_abundance_summary'\n47. 'plot_c2c_cells'\n48. 'plot_c2c_overview'\n49. 'plot_context_boxplot'\n50. 'plot_context_heat'\n51. 'plot_lr_heatmap'\n52. 'rank_aggregate'\n53. 'rank_method'\n54. 'recomplexify'\n55. 'sca_formatDB'\n56. 'select_resource'\n57. 'show_homologene'\n58. 'show_methods'\n59. 'show_resources'\n\n\n","text/latex":"\\begin{enumerate*}\n\\item 'assign\\_lr\\_weights'\n\\item 'calculate\\_gini'\n\\item 'call\\_cellchat'\n\\item 'call\\_connectome'\n\\item 'call\\_italk'\n\\item 'call\\_natmi'\n\\item 'call\\_sca'\n\\item 'call\\_squidpy'\n\\item 'cellchat\\_formatDB'\n\\item 'chord\\_freq'\n\\item 'compile\\_ligrec'\n\\item 'conn\\_formatDB'\n\\item 'decomplexify'\n\\item 'decompose\\_tensor'\n\\item 'filter\\_nonabundant\\_celltypes'\n\\item 'FormatConnectome'\n\\item 'FormatiTALK'\n\\item 'FormatSCA'\n\\item 'generate\\_homologs'\n\\item 'generate\\_lr\\_geneset'\n\\item 'generate\\_omni'\n\\item 'generate\\_orthologs'\n\\item 'get\\_abundance\\_summary'\n\\item 'get\\_c2c\\_factors'\n\\item 'get\\_connectome'\n\\item 'get\\_curated\\_omni'\n\\item 'get\\_logfc'\n\\item 'get\\_lr\\_resources'\n\\item 'get\\_natmi'\n\\item 'get\\_sca'\n\\item 'heat\\_freq'\n\\item 'italk\\_formatDB'\n\\item 'join\\_means'\n\\item 'liana\\_aggregate'\n\\item 'liana\\_bysample'\n\\item 'liana\\_call'\n\\item 'liana\\_defaults'\n\\item 'liana\\_dotplot'\n\\item 'liana\\_heatmap'\n\\item 'liana\\_pipe'\n\\item 'liana\\_prep'\n\\item 'liana\\_tensor\\_c2c'\n\\item 'liana\\_wrap'\n\\item 'mean0'\n\\item 'minmax'\n\\item 'plot\\_abundance\\_summary'\n\\item 'plot\\_c2c\\_cells'\n\\item 'plot\\_c2c\\_overview'\n\\item 'plot\\_context\\_boxplot'\n\\item 'plot\\_context\\_heat'\n\\item 'plot\\_lr\\_heatmap'\n\\item 'rank\\_aggregate'\n\\item 'rank\\_method'\n\\item 'recomplexify'\n\\item 'sca\\_formatDB'\n\\item 'select\\_resource'\n\\item 'show\\_homologene'\n\\item 'show\\_methods'\n\\item 'show\\_resources'\n\\end{enumerate*}\n","text/plain":[" [1] \"assign_lr_weights\"            \"calculate_gini\"              \n"," [3] \"call_cellchat\"                \"call_connectome\"             \n"," [5] \"call_italk\"                   \"call_natmi\"                  \n"," [7] \"call_sca\"                     \"call_squidpy\"                \n"," [9] \"cellchat_formatDB\"            \"chord_freq\"                  \n","[11] \"compile_ligrec\"               \"conn_formatDB\"               \n","[13] \"decomplexify\"                 \"decompose_tensor\"            \n","[15] \"filter_nonabundant_celltypes\" \"FormatConnectome\"            \n","[17] \"FormatiTALK\"                  \"FormatSCA\"                   \n","[19] \"generate_homologs\"            \"generate_lr_geneset\"         \n","[21] \"generate_omni\"                \"generate_orthologs\"          \n","[23] \"get_abundance_summary\"        \"get_c2c_factors\"             \n","[25] \"get_connectome\"               \"get_curated_omni\"            \n","[27] \"get_logfc\"                    \"get_lr_resources\"            \n","[29] \"get_natmi\"                    \"get_sca\"                     \n","[31] \"heat_freq\"                    \"italk_formatDB\"              \n","[33] \"join_means\"                   \"liana_aggregate\"             \n","[35] \"liana_bysample\"               \"liana_call\"                  \n","[37] \"liana_defaults\"               \"liana_dotplot\"               \n","[39] \"liana_heatmap\"                \"liana_pipe\"                  \n","[41] \"liana_prep\"                   \"liana_tensor_c2c\"            \n","[43] \"liana_wrap\"                   \"mean0\"                       \n","[45] \"minmax\"                       \"plot_abundance_summary\"      \n","[47] \"plot_c2c_cells\"               \"plot_c2c_overview\"           \n","[49] \"plot_context_boxplot\"         \"plot_context_heat\"           \n","[51] \"plot_lr_heatmap\"              \"rank_aggregate\"              \n","[53] \"rank_method\"                  \"recomplexify\"                \n","[55] \"sca_formatDB\"                 \"select_resource\"             \n","[57] \"show_homologene\"              \"show_methods\"                \n","[59] \"show_resources\"              "]},"metadata":{}}],"source":["ls(\"package:liana\")"]}],"metadata":{"colab":{"provenance":[],"collapsed_sections":["vjg2X61cpzdh"]},"kernelspec":{"display_name":"R","name":"ir"},"language_info":{"name":"R"}},"nbformat":4,"nbformat_minor":0}